src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter.cs

```diff
@@ -0,0 +1,106 @@
+            }
+
+            public override SyntaxNode VisitCasePatternSwitchLabel(CasePatternSwitchLabelSyntax node)
+            {
+                return node.Pattern;
+            }
+
+            public override SyntaxNode VisitCaseSwitchLabel(CaseSwitchLabelSyntax node)
+            {
+                return ConstantPattern(node.Value);
+            }
+
+            public override SyntaxNode VisitDefaultSwitchLabel(DefaultSwitchLabelSyntax node)
+            {
+                return DiscardPattern();
+            }
+
+            public override SyntaxNode VisitSwitchSection(SwitchSectionSyntax node)
+            {
+                Debug.Assert(node.Labels.Count == 1);
+                var switchLabel = node.Labels[0];
+                return SwitchExpressionArm(
+                    pattern: (PatternSyntax)Visit(switchLabel),
+                    whenClause: switchLabel.IsKind(SyntaxKind.CasePatternSwitchLabel, out CasePatternSwitchLabelSyntax patternSwitchLabel)
+                        ? patternSwitchLabel.WhenClause
+                        : null,
+                    expression: VisitStatements(node.Statements));
+            }
+
+            public override SyntaxNode VisitAssignmentExpression(AssignmentExpressionSyntax node)
+            {
+                (_assignmentTargetsOpt ?? (_assignmentTargetsOpt = new List<ExpressionSyntax>())).Add(node.Left);
+                return node.Right;
+            }
+
+            private ExpressionSyntax VisitStatements(SyntaxList<StatementSyntax> statements)
+            {
+                Debug.Assert(statements.Count > 0);
+                Debug.Assert(!statements[0].IsKind(SyntaxKind.BreakStatement));
+
+                var nodes = new List<ExpressionSyntax>();
+                foreach (var statement in statements)
+                {
+                    if (statement.IsKind(SyntaxKind.BreakStatement))
+                    {
+                        break;
+                    }
+
+                    nodes.Add((ExpressionSyntax)Visit(statement));
+                }
+
+                Debug.Assert(nodes.Count > 0);
+                return nodes.Count == 1 ? nodes[0] : TupleExpression(SeparatedList(nodes.Select(Argument)));
+            }
+
+            public override SyntaxNode VisitSwitchStatement(SwitchStatementSyntax node)
+            {
+                return SwitchExpression(node.Expression, SeparatedList(node.Sections.Select(VisitSwitchSection)));
+            }
+
+            public override SyntaxNode VisitReturnStatement(ReturnStatementSyntax node)
+            {
+                return node.Expression;
+            }
+
+            public override SyntaxNode VisitThrowStatement(ThrowStatementSyntax node)
+            {
+                return ThrowExpression(node.Expression);
+            }
+
+            public override SyntaxNode DefaultVisit(SyntaxNode node)
+            {
+                throw ExceptionUtilities.UnexpectedValue(node.Kind());
```

Comment:
This part really worries me.  due to the mismatch between iop and syntax i'm veyr wary of being able to make the claim that you won't hit this.   I'll have to think about this a bit.  But my gut makes me feel we should not assume this, but instead just set a bool or something if we ran into something we didn't understand.  And the analyzer should acutally try doing the rewrite, and should only report the diagnostic if the rewrite would succeed...

Reply:
been thinking about this. in fact, IOperation doesn't make things easier, specially if we want to support artbriray assignment targets. I've rewritten the analyzer to operate on syntax nodes.